lib/vector-core/buffers/src/lib.rs

```diff
@@ -205,3 +223,42 @@ impl<T, S: Sink<T> + Unpin> Sink<T> for DropWhenFull<S> {
         self.project().inner.poll_close(cx)
     }
 }
+
+// The BlockWhenFull wrapper is used for instrumentation purposes
+#[pin_project]
+pub struct BlockWhenFull<S> {
+    #[pin]
+    inner: S,
+}
+
+impl<S> BlockWhenFull<S> {
+    pub fn new(inner: S) -> Self {
+        Self { inner }
+    }
+}
+
+impl<T, S: Sink<T> + Unpin> Sink<T> for BlockWhenFull<S> {
+    type Error = S::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.project().inner.poll_ready(cx)
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        let byte_size = size_of_val(&item);
```

Comment:
And here as well, I think we'll want to differ this depending on whether it is a memory or disk buffer. For memory buffers, use `allocated_bytes`, for disk buffers, calculate the encoded size.